### PR TITLE
Feature: Implement the policy report field delete flow

### PR DESCRIPTION
**Pull request by waterim**
### Details

Added Delete report field  

### Fixed Issues

$ https://github.com/Expensify/App/issues/43686
$ https://github.com/Expensify/App/issues/44797
$ https://github.com/Expensify/App/issues/44798
PROPOSAL: N/A

### Tests

1. Navigate to Report fields in Workspace
2. Click on the Report field
3. Observe delete button(Note that delete button is in the correct place, it shouldnt be in 3dot)
4. Try to delete 

____

1. Go to staging.new.expensify.com
2. Go to Workspace settings > Report fields (enable from More features)
3. Go offline.
4. Tap Add field.
5. Add any report field and save it.
6. Verify the row will be grayed out when added offline.

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as tests

### QA Steps

None

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [x] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Web</summary>
image
image

</details>

### CODE REVIEW

**Comment by waterim** [comment]
@shubham1206agra fixed

**Comment by waterim** [comment]
@shubham1206agra fixed

File: src/pages/workspace/reportFields/WorkspaceReportFieldSettingsPage.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        ReportField.deletePolicyReportFields(route.params.policyID, [reportFieldKey]);
+        setIsDeleteModalVisible(false);
+        Navigation.goBack();
+    };
+
+    return (
+        <AccessOrNotFoundWrapper
+            accessVariants={[CONST.POLICY.ACCESS_VARIANTS.ADMIN, CONST.POLICY.ACCESS_VARIANTS.PAID]}
+            policyID={route.params.policyID}
+            featureName={CONST.POLICY.MORE_FEATURES.ARE_TAGS_ENABLED}
```
[the remark this answers is not in the document]
Comment (waterim) [reply src/pages/workspace/reportFields/WorkspaceReportFieldSettingsPage.tsx]: Thanks, fixed
Forgot to change it when more features was merged

**Comment by waterim** [comment]
@shubham1206agra Fixed and added tests

**Comment by waterim** [comment]
I see, thats because of the empty report Field in onyx, couldn't fix it fast today, will figure it out tomorrow morning!

**Comment by mountiny** [comment]
@waterim conflicts

File: src/libs/actions/Policy/ReportField.ts [hunk 2]
```diff
@@ -30,12 +30,11 @@ Onyx.connect({
 const allPolicies: OnyxCollection<Policy> = {};
 Onyx.connect({
     key: ONYXKEYS.COLLECTION.POLICY,
-    callback: (value, key) => {
+    callback: (val, key) => {
```
Comment (rezkiy37) [inline src/libs/actions/Policy/ReportField.ts:33]:
```suggestion
    callback: (value, key) => {
```

File: tests/actions/ReportFieldsTest.ts [hunk 3]
```diff
@@ -234,4 +235,73 @@ describe('actions/ReportFields', () => {
...
+            // check if the report field exists in the policy
+            expect(policy?.fieldList).toStrictEqual({
+                [reportFieldKey]: fakeReportField,
+            });
+
+            ReportField.deletePolicyReportFields(fakePolicy.id, [reportFieldKey]);
+            await waitForBatchedUpdates();
+
+            // Check for success data
+            (fetch as MockFetch)?.resume?.();
```
Comment (rezkiy37) [inline tests/actions/ReportFieldsTest.ts:284]:
```suggestion
            mockFetch?.resume?.();
```

File: tests/actions/ReportFieldsTest.ts [hunk 4]
```diff
@@ -234,4 +235,73 @@ describe('actions/ReportFields', () => {
...
+            await waitForBatchedUpdates();
+
+            policy = await new Promise((resolve) => {
+                const connectionID = Onyx.connect({
+                    key: ONYXKEYS.COLLECTION.POLICY,
```
Comment (rezkiy37) [inline tests/actions/ReportFieldsTest.ts:289]:
```suggestion
                   key: `${ONYXKEYS.COLLECTION.POLICY}${policyID}`,
```

File: tests/actions/ReportFieldsTest.ts [hunk 5]
```diff
@@ -234,4 +235,73 @@ describe('actions/ReportFields', () => {
             expect(policy?.pendingFields?.[reportFieldKey]).toBeFalsy();
         });
     });
+
+    describe('deleteReportField', () => {
+        it('Deleted a report field from a workspace', async () => {
```
Comment (rezkiy37) [inline tests/actions/ReportFieldsTest.ts:240]: add a failure case

**Review by rezkiy37 — CHANGES_REQUESTED** [review]
Minor comments.

File: src/libs/actions/Policy/ReportField.ts [hunk 6]
```diff
@@ -204,6 +199,76 @@ function createReportField(policyID: string, {name, type, initialValue}: CreateR
     API.write(WRITE_COMMANDS.CREATE_WORKSPACE_REPORT_FIELD, parameters, onyxData);
 }
 
+function deletePolicyReportFields(policyID: string, reportFieldsToUpdate: string[]) {
```
Comment (rezkiy37) [inline src/libs/actions/Policy/ReportField.ts:202]:
```suggestion
function deleteReportFields(policyID: string, reportFieldsToUpdate: string[]) {
```

**Review by shubham1206agra — APPROVED** [review]

File: src/libs/API/types.ts [hunk 7]
```diff
@@ -131,6 +131,7 @@ const WRITE_COMMANDS = {
     RENAME_POLICY_TAG: 'RenamePolicyTag',
     SET_WORKSPACE_REQUIRES_CATEGORY: 'SetWorkspaceRequiresCategory',
     DELETE_WORKSPACE_CATEGORIES: 'DeleteWorkspaceCategories',
+    POLICY_REPORT_FIELDS_REPLACE: 'Policy_ReportFields_Replace',
```
Comment (mountiny) [inline src/libs/API/types.ts:134]: @waterim we should be using `DeletePolicyReportField`, its the same but we need to update this

File: src/SCREENS.ts [hunk 8]
```diff
@@ -318,6 +318,7 @@ const SCREENS = {
         TAG_EDIT: 'Tag_Edit',
         TAXES: 'Workspace_Taxes',
         REPORT_FIELDS: 'Workspace_ReportFields',
+        REPORT_FIELD_SETTINGS: 'Workspace_ReportField_Settings',
```
Comment (mountiny) [inline src/SCREENS.ts:321]: I think we should keep this cohesive and keep all the names same so.
```suggestion
        REPORT_FIELDS_SETTINGS: 'Workspace_ReportFields_Settings',
```

If there would be other page that fits `settings` more we can rename this to `details`

**Review by mountiny — APPROVED** [review]
Couple changes that are I think only NAB for this PR but should be addressed in a follow up @waterim going to merge, thank you

File: src/SCREENS.ts (hunk 8 above)
Comment (waterim) [reply src/SCREENS.ts:321]: @mountiny This is a different page, we have a report_FIELD_settings -> settings for a single report field
And we have report_FIELDS_settings -> settings for all report fields

**Pull request merged**